### PR TITLE
Refactor how model configs are combined

### PR DESCRIPTION
**Pull request by mwhudson**
Basically each model now produces a fragment of a complete curtin config
and they are combined using curtin's merge_config function rather than
SubiquityModel.render knowing the shape of the config produced by each
model.

I've been meaning to do this for a while. It would be easy to mess things up here so please review with a bit of care.

### CODE REVIEW

**Comment by xnox** [comment]
Note to self:
The interesting pieces to validate are that e.g. `write_files` is correctly merged across all modules and such that e.g. setting proxy doesn't remove the keyboard write_files snipper. Or that apt config has both mirror and proxy. Or that maas install path write_files are still there.

**Comment by mwhudson** [comment]
Yeah, that part is fine: the reason that curtin config has dictionaries in places you might expect lists (like write_files) is exactly so merge_configs dtrt when merging them (or at least, I assume that this is the reason, I haven't asked anyone).

**Comment by raharper** [comment]
I would suggest adding a unittest on this path to ensure that newly added snippets don't clobber things, but generally looking at:
https://git.launchpad.net/curtin/tree/tests/unittests/test_config.py

Should demonstrate expected merging behavior of config.merge_config.

Note, curtin does not have an advanced config for controllering *how* things are merged.
So, generally, "last" key/value pair wins, much like dict.update().

File: subiquity/models/subiquity.py [hunk 1]
```diff
@@ -68,18 +74,30 @@ def __init__(self, common):
...
+
+        # Collect the models that produce data for the curtin config.
+        self._install_models = [
+            self.keyboard,
+            self.installpath,
+            self.network,
+            self.proxy,
+            self.mirror,
+            self.filesystem,
+            ]
```
Comment (raharper) [inline subiquity/models/subiquity.py]: this list seems brittle, is it possible to just call render() on all models?  If an new or existing model adds curtin config, then this list needs updating.

Comment (mwhudson) [reply subiquity/models/subiquity.py]: I sort of agree but there's no simple way to get a list of "all models" either really?

**Comment by mwhudson** [comment]
> I would suggest adding a unittest on this path to ensure that newly added snippets don't clobber things

This was a good idea, see what you think of the tests I've added now.

File: subiquity/models/subiquity.py (hunk 1 above)
Comment (xnox) [reply subiquity/models/subiquity.py]: One could interate the attributes of the object to see which ones are models.... but at the moment we'd need to like do a fuzzy match that `__class__.endswith('Model')` as all of them are just "objects" and don't inherit from anything. Maybe they should inherit all from a base-class Model, that at least implements `def render(): pass` but meh.

File: subiquity/models/installpath.py [hunk 2]
```diff
@@ -38,26 +38,30 @@ def paths(self):
...
             cmdline = [(_('Install from cli provided sources'), 'cmdline')]
         return cmdline + [
-            (_('Install Ubuntu'),                 'ubuntu'),
+            (_('Install Ubuntu'),                         'ubuntu'),
             (_('Install MAAS bare-metal cloud (region)'), 'maas_region'),
             (_('Install MAAS bare-metal cloud (rack)'),   'maas_rack'),
         ]
 
-    @property
-    def sources(self):
```
Comment (xnox) [inline subiquity/models/installpath.py]: This removed property is referenced by `get_seed_yaml` here https://github.com/mwhudson/subiquity/blob/refactor-config-creation/subiquity/ui/views/snaplist.py#L415

Not sure if we want to keep the sources property, or make the snaplist reference and transverse the full rendered config instead?

Currently it bobms out with assertionerror. So can't merge this.

File: subiquitycore/i18n.py [hunk 3]
```diff
@@ -29,6 +29,7 @@
 
 
 def switch_language(code='en_US'):
+    syslog.syslog('switch_language ' + code)
```
Comment (raharper) [inline subiquitycore/i18n.py]: odd change for this PR

**Pull request merged**